common/hashing.h

```diff
@@ -0,0 +1,721 @@
+// a 64-bit multiply efficiently.
+//
+// The function supports many typical data types such as primitives, string-ish
+// views, and types composing primitives transparently like pairs, tuples, and
+// array-ish views. It is also extensible to support user-defined types.
+//
+// To add support for your type, you need to implement a customization point --
+// a free function that can be found by ADL for your type -- called `CarbonHash`
+// with the following signature:
+//
+// ```cpp
+// auto CarbonHash(HashState& h, const YourType& value) -> void;
+// ```
+//
+// See the comments on the `Hasher` type for more details about implementing
+// these customization points.
+template <typename T>
+inline auto HashValue(const T& value, uint64_t seed) -> HashCode;
+
+// The same as the seeded version of `HashValue` but without callers needing to
+// provide a seed.
+//
+// Generally prefer the seeded version, but this is available if there is no
+// reasonable seed. In particular, this will behave better than using a seed of
+// `0`.
+template <typename T>
+inline auto HashValue(const T& value) -> HashCode;
+
+// Object and APIs that eventually produce a hash code.
+//
+// This type is primarily used by types to implement a customization point
+// `CarbonHash` that will in turn be used by the `HashValue` function. See the
+// `HashValue` function for details of that extension point.
+//
+// The API of this type can be used to incorporate data from your user-defined
+// type into a hash code. The API is a low-level collection of primitives rather
+// than a higher-level abstraction. This reflects a fundamental priority: deep,
+// detailed, and latency-oriented optimization of any hash functions. This
+// type's API also reflects the reality that high-performance hash tables rely
+// on keys that are generally small and cheap to hash. The result is
+// prioritizing hashing a small number of integers (1 or 2 ideally), or some
+// contiguous byte buffer. We typically want hash tables whose key types work to
+// fit into this model rather than advanced or more complex hashing of more
+// complex input data.
+//
+// It also results in an unusual API design where all the functions are actually
+// *static* member functions that expect the `Hasher` and its state to be moved
+// into the first parameter and returned, updated, by value. This is in contrast
+// to non-static member functions in C++ which would mutate the state in place
+// by taking its address for the implicit `this` parameter. While that *usually*
+// optimizes away, it creates a more difficult optimization environment that has
+// resulted in performance issues over time and at scale.
+//
+// Another important note is that types only need to include the data that would
+// lead to an object comparing equal or not-equal to some other object,
+// including objects of other types if a type supports heterogeneous equality
+// comparison.
```

Comment:
IIUC the current handling of negative integers doesn't satisfy this requirement -- `int32_t(-1)` and `int64_t(-1)` will have different hash values, even though they compare equal.